graphrag/index/run/run.py

```diff
@@ -0,0 +1,265 @@
+            - id - The id of the document
+            - text - The text of the document
+            - title - The title of the document
+            These must exist after any post process steps are run if there are any!
+        - storage - The storage to use for the pipeline
+        - cache - The cache to use for the pipeline
+        - reporter - The reporter to use for the pipeline
+        - input_post_process_steps - The post process steps to run on the input data
+        - additional_verbs - The custom verbs to use for the pipeline
+        - additional_workflows - The custom workflows to use for the pipeline
+        - debug - Whether or not to run in debug mode
+    Returns:
+        - output - An iterable of workflow results as they complete running, as well as any errors that occur
+    """
+    start_time = time.time()
+    stats = PipelineRunStats()
+    storage = storage or MemoryPipelineStorage()
+    cache = cache or InMemoryCache()
+    progress_reporter = progress_reporter or NullProgressReporter()
+    callbacks = callbacks or ConsoleWorkflowCallbacks()
+    callbacks = _create_callback_chain(callbacks, progress_reporter)
+    emit = emit or [TableEmitterType.Parquet]
+    emitters = create_table_emitters(
+        emit,
+        storage,
+        lambda e, s, d: cast(WorkflowCallbacks, callbacks).on_error(
+            "Error emitting table", e, s, d
+        ),
+    )
+    loaded_workflows = load_workflows(
+        workflows,
+        additional_verbs=additional_verbs,
+        additional_workflows=additional_workflows,
+        memory_profile=memory_profile,
+    )
+    workflows_to_run = loaded_workflows.workflows
+    workflow_dependencies = loaded_workflows.dependencies
+
+    context = _create_run_context(storage, cache, stats)
+
+    if len(emitters) == 0:
+        log.info(
+            "No emitters provided. No table outputs will be generated. This is probably not correct."
+        )
+
+    dataset = await _run_post_process_steps(
+        input_post_process_steps, dataset, context, callbacks
+    )
+
+    # Make sure the incoming data is valid
+    _validate_dataset(dataset)
+
+    log.info("Final # of rows loaded: %s", len(dataset))
+    stats.num_documents = len(dataset)
+    last_workflow = "input"
+
+    try:
+        await _dump_stats(stats, storage)
+
+        for workflow_to_run in workflows_to_run:
+            # Try to flush out any intermediate dataframes
+            gc.collect()
+
+            workflow = workflow_to_run.workflow
+            workflow_name: str = workflow.name
+            last_workflow = workflow_name
+
+            log.info("Running workflow: %s...", workflow_name)
+
+            if is_resume_run and await storage.has(
+                f"{workflow_to_run.workflow.name}.parquet"
+            ):
+                log.info("Skipping %s because it already exists", workflow_name)
+                continue
+
+            stats.workflows[workflow_name] = {"overall": 0.0}
```

Comment:
Probably not part of this PR but stats.workflows could be a default dict if it is always initialized this way

Reply:
I like this idea!